### ISSUE TITLE
rustc_args_required_const should mark its const arguments as const

### ISSUE THREAD

**Issue by lu-zero** [rust-lang/rust#61180]
calling a `rustc_args_required_const`-marked function from a `rustc_args_required_const`-marked function does not work as intended.

[here the playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=2be212c8ee1c151ed5fcdfa2dc71c89d)
``` rust
#![feature(rustc_attrs)]

trait T {
    #[rustc_args_required_const(1)]
    fn t(self, b: i32);
}

#[inline]
#[rustc_args_required_const(1)]
const fn ex_t(a: u8, b: i32) {
    let _ = a as i32 + b;
}

#[inline]
#[rustc_args_required_const(1)]
fn ex_t2(a: u8, b: i32) {
    ex_t(a, b);
}

impl T for u8 {
    #[rustc_args_required_const(1)]
    fn t(self, b: i32) {
        // println!("{}", b);
        ex_t(self, b);
    }
}

fn main() {
    let a = 42u8;
    
    a.t(42);
}
```

**Comment by gnzlbg** [comment]
> calling a rustc_args_required_const-marked function from a rustc_args_required_const-marked function does not work as intended.

That is not intended to work, so I would close this issue.

**Comment by gnzlbg** [comment]
`#[rustc_args_required_const(N)]` just errors if the `N` argument is not const. That's it. It is not any sort of magical type-level consts in argument position kind of thing. 

You can turn a run-time value into a compile-time constant by, e.g., matching on it, e.g., using `match` like this: `match N { 0 => 0, 1 => 1, ... , _ => 255, }`.

**Comment by oli-obk** [comment]
Some discussion on zulip on this: https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/rustc_args_required_const.20in.20traits

TLDR: I don't think we should add any more features to `rustc_args_required_const`. We should either go with const generics or have an actual RFC that figures out how to have const generics syntactic sugar in argument position

**Comment by Mark-Simulacrum** [comment]
Closing as per the above comments.
